File: sample/Sample-OC/ViewController.m
//
//  ViewController.m
//  Sample-OC
//
//  Created by CavanSu on 2020/8/18.
//  Copyright © 2020 CavanSu. All rights reserved.
//

#import "ViewController.h"
#import <YKNet/YKNet-Swift.h>

@interface ViewController () <YKNetDelegateOC, YKNetLogTubeOC>
@property (nonatomic, strong) YKNetOC *client;
@end

@implementation ViewController
- (void)viewDidLoad {
    [super viewDidLoad];
    self.client = [[YKNetOC alloc] initWithDelegate:self
                                            logTube:self];
    [self getRequest];
}

- (void)getRequest {
    NSString *url = @"https://www.tianqiapi.com/api";
    YKNetRequestEventOC *event = [[YKNetRequestEventOC alloc] initWithName:@"Sample-Get"];
    YKNetRequestTypeObjectOC *type = [[YKNetRequestTypeJsonObjectOC alloc] initWithMethod:YKNetHTTPMethodOCGet
                                                                                url:url];
    
    YKNetRequestTaskOC *task = [[YKNetRequestTaskOC alloc] initWithEvent:event
                                                              type:type
                                                           timeout:10
                                                            header:nil
                                                        parameters:@{@"appid": @"23035354",
                                                                     @"appsecret": @"8YvlPNrz",
                                                                     @"version": @"v9",
                                                                     @"cityid": @"0",
                                                                     @"city": @"%E9%9D%92%E5%B2%9B",
                                                                     @"ip": @"0",
                                                                     @"callback": @"0"}];
    
    [self.client requestWithTask:task
             responseOnQueue:nil
          successCallbackContent:YKNetResponseTypeOCJson
                         success:^(YKNetResponseOC * _Nonnull response) {
        NSLog(@"weather json: %@", response.json);
    } fail:^NSTimeInterval(YKNetErrorOC * _Nonnull error) {
        NSLog(@"error: %@", error.localizedDescription);
        return 0;
    }];
}

- (void)postRequest {
    NSString *url = @"";
    YKNetRequestEventOC *event = [[YKNetRequestEventOC alloc] initWithName:@"Sample-Post"];
    YKNetRequestTypeObjectOC *type = [[YKNetRequestTypeJsonObjectOC alloc] initWithMethod:YKNetHTTPMethodOCPost
                                                                                url:url];
    
    YKNetRequestTaskOC *task = [[YKNetRequestTaskOC alloc] initWithEvent:event
                                                              type:type
                                                           timeout:10
                                                            header:nil
                                                        parameters:nil];
    
    [self.client requestWithTask:task
                 responseOnQueue:nil
          successCallbackContent:YKNetResponseTypeOCJson
                         success:^(YKNetResponseOC * _Nonnull response) {
        NSLog(@"weather json: %@", response.json);
    } fail:^NSTimeInterval(YKNetErrorOC * _Nonnull error) {
        NSLog(@"error: %@", error.localizedDescription);
        return 0;
    }];
}

- (void)uploadTask {
    NSString *url = @"";
    YKNetRequestEventOC *event = [[YKNetRequestEventOC alloc] initWithName:@"Sample-Upload"];
    YKNetUploadObjectOC *object = [[YKNetUploadObjectOC alloc] initWithFileKeyOnServer:@"server-input"
                                                                        fileName:@"test"
                                                                        fileData:[[NSData alloc] init]
                                                                            mime:YKNetFileMIMEOCPng];
    
    YKNetUploadTaskOC *task = [[YKNetUploadTaskOC alloc] initWithEvent:event
                                                         timeout:10
                                                          object:object
                                                             url:url
                                                          header:nil
                                                      parameters:nil];
    
    [self.client uploadWithTask:task
                responseOnQueue:nil
         successCallbackContent:YKNetResponseTypeOCBlank
                        success:^(YKNetResponseOC * _Nonnull response) {
        NSLog(@"upload success");
    } fail:^NSTimeInterval(YKNetErrorOC * _Nonnull error) {
        NSLog(@"error: %@", error.localizedDescription);
        return -1;
    }];
}

- (void)downloadTask {
    NSString *url = @"";
    YKNetRequestEventOC *event = [[YKNetRequestEventOC alloc] initWithName:@"Sample-Download"];
    YKNetDownloadObjectOC *object = [[YKNetDownloadObjectOC alloc] initWithTargetDirectory:@""
                                                                               cover:YES];
    YKNetDownloadTaskOC *task = [[YKNetDownloadTaskOC alloc] initWithEvent:event
                                                             timeout:10
                                                              object:object
                                                                 url:url
                                                              header:nil
                                                          parameters:nil];
    
    [self.client downloadWithTask:task
                  responseOnQueue:nil
           successCallbackContent:YKNetResponseTypeOCJson
                         progress:^(float progress) {
        NSLog(@"%f",progress);
    } success:^(YKNetResponseOC * response) {
        NSLog(@"download success");
    } fail:^NSTimeInterval(YKNetErrorOC * error) {
        NSLog(@"error: %@", error.localizedDescription);
        return -1;
    }];
}

#pragma mark - YKNetDelegateOC, ArLogTube
- (void)ykNet:(YKNetOC *)client
requestSuccess:(YKNetRequestEventOC *)event
    startTime:(NSTimeInterval)startTime
          url:(NSString *)url {
    NSLog(@"event: %@, requestSuccess, url: %@", event.name, url);
}

- (void)ykNet:(YKNetOC *)client
  requestFail:(YKNetErrorOC *)error
        event:(YKNetRequestEventOC *)event
          url:(NSString *)url {
    NSLog(@"event: %@, requestFail, url: %@", event.name, url);
}

- (void)logWithInfo:(NSString *)info
              extra:(NSString *)extra {
    NSLog(@"log info: %@, extra: %@", info, extra);
}

- (void)logWithWarning:(NSString *)warning
                 extra:(NSString *)extra {
    NSLog(@"log warning: %@, extra: %@", warning, extra);
}

- (void)logWithError:(YKNetErrorOC *)error
               extra:(NSString *)extra {
    NSLog(@"log error: %@, extra: %@", error, extra);
}
@end

File: sample/Sample/ViewController.swift
//
//  ViewController.swift
//  Sample
//
//  Created by CallmeLetty on 2020/5/25.
//  Copyright © 2020 CallmeLetty. All rights reserved.
//

import UIKit
import YKNet

class ViewController: UIViewController {
    lazy var client = YKNet(delegate: self,
                            logTube: self)
    
    override func viewDidLoad() {
        super.viewDidLoad()
    }
    
    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        getRequest()
    }
    
    func getRequest() {
        let url = "https://www.tianqiapi.com/api"
        let event = YKNetRequestEvent(name: "Sample-get")
        let task = YKNetRequestTask(event: event,
                                 type: .http(.get, url: url),
                                 timeout: .low,
                                 parameters: ["appid": "23035354",
                                              "appsecret": "8YvlPNrz",
                                              "version": "v9",
                                              "cityid": "0",
                                              "city": "%E9%9D%92%E5%B2%9B",
                                              "ip": "0",
                                              "callback": "0"])
        
        client.request(task: task, success: YKNetResponse.json({ (json) in
            print("weather json: \(json.description)")
        })) { (error) -> YKNetRetryOptions in
            print("error: \(error.localizedDescription)")
            return .resign
        }
    }
    
    func postRequest() {
        let url = ""
        let event = YKNetRequestEvent(name: "Sample-post")
        let parameters: [String: Any]? = nil
        let headers: [String: String]? = nil
        let task = YKNetRequestTask(event: event,
                               type: .http(.post, url: url),
                               timeout: .low,
                               header: headers,
                               parameters: parameters)
        
        client.request(task: task, success: YKNetResponse.json({ (json) in
            print("weather json: \(json.description)")
        })) { (error) -> YKNetRetryOptions in
            print("error: \(error.localizedDescription)")
            return .resign
        }
    }
    
    func uploadRequest() {
        
    }
    
    func downloadRequest() {
        let urlStr = "https://convertcdn.netless.link/publicFiles.zip"
        let event = YKNetRequestEvent(name: "download-image")
        let obj = YKNetDownloadObject(targetDirectory: "/Users/doublecircle/Desktop/YKNetmin_local")
        let task = YKNetDownloadTask(event: event,
                                  object: obj,
                                  url: urlStr)
        client.download(task: task,
                        progress: { progress in
            print("----------Progress: \(progress)")
        }, success: YKNetResponse.string({ finalPath in
            print("----final path: \(finalPath)")
        })) { (error) -> YKNetRetryOptions in
            print("error: \(error.localizedDescription)")
            return .resign
        }
    }
}

// MARK: - YKNetDelegate
extension ViewController: YKNetDelegate {
    func ykNet(_ client: YKNet,
               requestSuccess event: YKNetRequestEvent,
               startTime: TimeInterval, url: String) {
        print("request success, event: \(event.description), url: \(url)")
    }
    
    func ykNet(_ client: YKNet,
               requestFail error: YKNetError,
               event: YKNetRequestEvent, url: String) {
        print("request error, event: \(event.description), error: \(error.localizedDescription), url: \(url)")
    }
}

// MARK: - YKNetLogTube
extension ViewController: YKNetLogTube {
    func log(info: String,
             extra: String?) {
        print("info: \(info), extra: \(extra ?? "nil")")
    }
    
    func log(warning: String,
             extra: String?) {
        print("warning: \(warning), extra: \(extra ?? "nil")")
    }
    
    func log(error: YKNetError, extra: String?) {
        print("error: \(error), extra: \(extra ?? "nil")")
    }
}
